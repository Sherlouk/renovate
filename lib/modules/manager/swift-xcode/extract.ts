import semverInc from 'semver/functions/inc';
import { logger } from '../../../logger';
import { getSiblingFileName, localPathExists } from '../../../util/fs';
import { regEx } from '../../../util/regex';
import { GitTagsDatasource } from '../../datasource/git-tags';
import type { PackageDependency, PackageFile } from '../types';

const regExps = {
  entireSection: regEx(
    /\/\* Begin XCRemoteSwiftPackageReference section \*\/([\0-\uFFFF]*)\/\* End XCRemoteSwiftPackageReference section \*\//
  ),
  eachPackage: regEx(
    /XCRemoteSwiftPackageReference "([^"]+?)".*?{(.*?};)(\t{2}|\s+)};/g
  ),
  packageIsa: regEx(/isa = XCRemoteSwiftPackageReference/),
  packageUrl: regEx(/repositoryURL = "(.*?)"/),
  packageRequirementKind: regEx(/kind = (.+?);/),
  packageRequirementMinimum: regEx(/minimumVersion = (.+?);/),
  packageRequirementMaximum: regEx(/maximumVersion = (.+?);/),
  packageRequirementVersion: regEx(/version = (.+?);/),
  packageRequirementBranch: regEx(/branch = (.+?);/),
  packageRequirementRevision: regEx(/revision = (.+?);/),
};

export async function extractPackageFile(
  content: string,
  packageFile: string = null
): Promise<PackageFile | null> {
  if (!content) {
    return null;
  }

  const sectionMatch = regExps.entireSection.exec(content);

  if (!sectionMatch || !sectionMatch[1]) {
    logger.debug(`${packageFile} contains no Swift package dependencies`);
    return null;
  }

  const matchText = sectionMatch[1].replaceAll('\n', '');
  const deps: PackageDependency[] = [];

  let packageMatch;
  while ((packageMatch = regExps.eachPackage.exec(matchText)) !== null) {
    const depName: string = packageMatch[1];

    if (!packageMatch[2].match(regExps.packageIsa)) {
      logger.debug(`${depName} has invalid ISA value`);
      continue;
    }

    let currentValue: string;
    let datasource: string;

    switch (packageMatch[2].match(regExps.packageRequirementKind)[1]) {
      case 'upToNextMinorVersion': {
        const minimum: string = packageMatch[2].match(
          regExps.packageRequirementMinimum
        )[1];
        const maximum = semverInc(minimum, 'minor');
        currentValue = `"${minimum}"..<"${maximum}"`;
        break;
      }
      case 'upToNextMajorVersion': {
        const minimum: string = packageMatch[2].match(
          regExps.packageRequirementMinimum
        )[1];
        const maximum = semverInc(minimum, 'major');
        currentValue = `"${minimum}"..<"${maximum}"`;
        break;
      }
      case 'versionRange': {
        const minimum: string = packageMatch[2].match(
          regExps.packageRequirementMinimum
        )[1];
        const maximum: string = packageMatch[2].match(
          regExps.packageRequirementMaximum
        )[1];
        currentValue = `"${minimum}"..<"${maximum}"`;
        break;
      }
      case 'exactVersion': {
        currentValue = packageMatch[2].match(
          regExps.packageRequirementVersion
        )[1];
        break;
      }
      case 'branch': {
        const value: string = packageMatch[2].match(
          regExps.packageRequirementBranch
        )[1];

        deps.push({
          currentValue: value,
          packageName: packageMatch[2].match(regExps.packageUrl)[1],
          depName,
          skipReason: 'unsupported-version',
        });
        continue;
      }
      case 'revision': {
        const value: string = packageMatch[2].match(
          regExps.packageRequirementRevision
        )[1];

        deps.push({
          currentValue: value,
          packageName: packageMatch[2].match(regExps.packageUrl)[1],
          depName,
          skipReason: 'is-pinned',
        });
        continue;
      }
    }

    if (!currentValue) {
      logger.debug(`${depName} has unknown revision requirements`);
      continue;
    }

    deps.push({
      currentValue: currentValue,
      datasource: datasource || GitTagsDatasource.id,
      packageName: packageMatch[2].match(regExps.packageUrl)[1],
      depName,
    });
  }

  if (!deps.length) {
    return null;
  }

  const res: PackageFile = { deps };
  const lockFile = getSiblingFileName(
    packageFile,
    'project.xcworkspace/xcshareddata/swiftpm/Package.resolved'
  );

  if (await localPathExists(lockFile)) {
    res.lockFiles = [lockFile];
  }

  return res;
}
