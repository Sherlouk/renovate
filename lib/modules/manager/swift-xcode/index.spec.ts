import { Fixtures } from '../../../../test/fixtures';
import { extractPackageFile } from './extract';

describe('modules/manager/swift-xcode/index', () => {
  describe('extractPackageFile()', () => {
    it('returns null for empty content', () => {
      expect(extractPackageFile(null)).toBeNull();
      expect(extractPackageFile(``)).toBeNull();
      expect(extractPackageFile(`// !$*UTF8*$!`)).toBeNull();
    });

    it('returns null for invalid content', () => {
      expect(
        extractPackageFile(`Begin XCRemoteSwiftPackageReference section`)
      ).toBeNull();
      expect(
        extractPackageFile(`/* Begin XCRemoteSwiftPackageReference section */`)
      ).toBeNull();
      expect(
        extractPackageFile(
          `/* Begin XCRemoteSwiftPackageReference section */ /* End XCRemoteSwiftPackageReference section */`
        )
      ).toBeNull();
    });

    it('returns null for invalid parameters', () => {
      expect(
        extractPackageFile(`Begin XCRemoteSwiftPackageReference section`)
      ).toBeNull();
      expect(
        extractPackageFile(`/* Begin XCRemoteSwiftPackageReference section */`)
      ).toBeNull();

      expect(
        extractPackageFile(
          // no package name
          `/* Begin XCRemoteSwiftPackageReference section */
          1671F0B028099F69001CAE81 /* XCRemoteSwiftPackageReference */ = {
            isa = XCRemoteSwiftPackageReference;
            repositoryURL = "https://github.com/pointfreeco/swift-snapshot-testing";
            requirement = {
              kind = upToNextMajorVersion;
              minimumVersion = 1.0.0;
            };
          };
        /* End XCRemoteSwiftPackageReference section */`
        )
      ).toBeNull();

      expect(
        extractPackageFile(
          // invalid ISA
          `/* Begin XCRemoteSwiftPackageReference section */
          1671F0B028099F69001CAE81 /* XCRemoteSwiftPackageReference */ = {
            isa = XCLocalSwiftPackageReference;
            repositoryURL = "https://github.com/pointfreeco/swift-snapshot-testing";
            requirement = {
              kind = upToNextMajorVersion;
              minimumVersion = 1.0.0;
            };
          };
        /* End XCRemoteSwiftPackageReference section */`
        )
      ).toBeNull();

      expect(
        extractPackageFile(
          // unknown requirement
          `/* Begin XCRemoteSwiftPackageReference section */
          1671F0B028099F69001CAE81 /* XCRemoteSwiftPackageReference */ = {
            isa = XCLocalSwiftPackageReference;
            repositoryURL = "https://github.com/pointfreeco/swift-snapshot-testing";
            requirement = {
              kind = upToNextPatchVersion;
              minimumVersion = 1.0.0;
            };
          };
        /* End XCRemoteSwiftPackageReference section */`
        )
      ).toBeNull();

      expect(
        extractPackageFile(
          // missing requirement
          `/* Begin XCRemoteSwiftPackageReference section */
          1671F0B028099F69001CAE81 /* XCRemoteSwiftPackageReference */ = {
            isa = XCLocalSwiftPackageReference;
            repositoryURL = "https://github.com/pointfreeco/swift-snapshot-testing";
          };
        /* End XCRemoteSwiftPackageReference section */`
        )
      ).toBeNull();
    });

    it('parses package descriptions', () => {
      expect(
        extractPackageFile(
          `/* Begin XCRemoteSwiftPackageReference section */
          1671F0B028099F69001CAE81 /* XCRemoteSwiftPackageReference "swift-snapshot-testing" */ = {
            isa = XCRemoteSwiftPackageReference;
            repositoryURL = "https://github.com/pointfreeco/swift-snapshot-testing";
            requirement = {
              kind = upToNextMajorVersion;
              minimumVersion = 1.0.0;
            };
          };
        /* End XCRemoteSwiftPackageReference section */`
        ).deps[0]
      ).toEqual(
        expect.objectContaining({
          packageName: 'swift-snapshot-testing',
          currentValue: '"1.0.0"..<"2.0.0"',
        })
      );
    });

    it('parses multiple packages', () => {
      expect(
        extractPackageFile(Fixtures.get(`project.pbxproj`))
      ).toMatchSnapshot();
    });
  });
});
