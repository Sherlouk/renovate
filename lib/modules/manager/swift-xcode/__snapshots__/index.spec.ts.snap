// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/swift-xcode/index extractPackageFile() parses multiple packages 1`] = `
Object {
  "deps": Array [
    Object {
      "currentValue": "\\"1.0.0\\"..<\\"2.0.0\\"",
      "datasource": "git-tags",
      "packageName": "swift-snapshot-testing",
      "sourceUrl": "https://github.com/pointfreeco/swift-snapshot-testing",
    },
    Object {
      "currentValue": "\\"5.6.0\\"..<\\"5.7.0\\"",
      "datasource": "git-tags",
      "packageName": "Alamofire",
      "sourceUrl": "https://github.com/Alamofire/Alamofire",
    },
    Object {
      "currentValue": "3.7.2",
      "datasource": "git-tags",
      "packageName": "CocoaLumberjack",
      "sourceUrl": "https://github.com/CocoaLumberjack/CocoaLumberjack",
    },
    Object {
      "currentValue": "master",
      "datasource": "git-tags",
      "packageName": "Kingfisher",
      "sourceUrl": "https://github.com/onevcat/Kingfisher",
    },
    Object {
      "currentValue": "6b0cda4ac8a5a8289afbacdb0cf428af2f3b9e30",
      "datasource": "git-tags",
      "packageName": "Nuke",
      "sourceUrl": "https://github.com/kean/Nuke",
    },
    Object {
      "currentValue": "\\"1.20.0\\"..<\\"1.25.0\\"",
      "datasource": "git-tags",
      "packageName": "SkeletonView",
      "sourceUrl": "https://github.com/Juanpe/SkeletonView",
    },
  ],
  "packageFile": null,
}
`;
